### DISCUSSION TITLE
Issues in in finding dseupd_c in c++ file

### DISCUSSION THREAD

**Discussion by Elil50** [opencollab/arpack-ng#447] [Q&A] [answered]
When i try to use dsaupd_c and dseupd_c from arpack-ng in a  c++ file, linking:

    #include "arpack.h"
    #include "arpackdef.h"

and linking these directories in the compilation process:

    /arpack-ng/SRC
    /arpack-ng/build
    /arpack-ng/ICB

I get this error, just for dseupd_c:

    OSError: ../file.so: undefined symbol: dseupd_c

If i use fortran, with a different program, it works fine. In that case i link this directory:

    /arpack-ng/build/libarpackILP64.a

I assumed the issue was the lack of ISO_C_BINDING compilation. What i usually do to compile arpack is:

Inside the arpack-ng folder, create a dir called build and move inside that folder:

    mkdir build
    cd build

Set the installation:

    cmake -D INTERFACE64=ON -D ITF64SUFFIX="ILP64" -D BUILD_SHARED_LIBS=ON -D BLA_VENDOR=Intel10_64ilp ..

Build the libraries by running just:

    make

and then install it by running:

    sudo make install

We have installed the shared libraries, we are going to need the static library also, so execute the following commands:

    cd ../
    rm -r -f build
    mkdir build
    cd build
    cmake -D INTERFACE64=ON -D ITF64SUFFIX="ILP64" -D BUILD_SHARED_LIBS=OFF -D BLA_VENDOR=Intel10_64ilp ..
    make
    sudo make install

And it works

However in the arpack-ng i find: [reminder 1]

    $ ./bootstrap
    $ export FFLAGS='-DMKL_ILP64 -I/usr/include/mkl'
    $ export FCFLAGS='-DMKL_ILP64 -I/usr/include/mkl'
    $ export LIBS='-Wl,--no-as-needed -L/usr/lib/x86_64-linux-gnu -lmkl_sequential -lmkl_core -lpthread -lm -ldl'
    $ export INTERFACE64=1
    $ ./configure --with-blas=mkl_gf_ilp64 --with-lapack=mkl_gf_ilp64
    $ make all check

And it gives me an error i will attach at the end.

In particular, if i run ./configure it says:

    checking for a BSD-compatible install... /usr/bin/install -c
    checking whether build environment is sane... yes
    checking for a race-free mkdir -p... /usr/bin/mkdir -p
    checking for gawk... gawk
    checking whether make sets $(MAKE)... yes
    checking whether make supports nested variables... yes
    checking whether to enable maintainer-specific portions of Makefiles... no
    checking for gfortran... no
    checking for g95... no
    checking for xlf95... no
    checking for f95... no
    checking for fort... no
    checking for ifort... ifort
    checking whether the Fortran compiler works... no
    configure: error: in `/arpack-ng':
    configure: error: Fortran compiler cannot create executables
    See `config.log' for more details

However when I use those compilers, they work.

If I run:

    ifort --version

It says:

    ifort (IFORT) 2021.11.0 20231010
    Copyright (C) 1985-2023 Intel Corporation.  All rights reserved.

And if I run:

    which ifort

It says:

    /opt/intel/oneapi/compiler/2024.0/bin/ifort

All these compiler issue don't appear if I just go for the steps i wrote at the beginning and run the program.

I tried to write [reminder 1] outside the build file i already made:

    autoreconf: export WARNINGS=
    autoreconf: Entering directory '.'
    autoreconf: configure.ac: not using Gettext
    autoreconf: running: aclocal --force -I m4
    autoreconf: configure.ac: tracing
    autoreconf: configure.ac: creating directory build-aux
    autoreconf: running: libtoolize --copy --force
    libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
    libtoolize: copying file 'build-aux/ltmain.sh'
    libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
    libtoolize: copying file 'm4/libtool.m4'
    libtoolize: copying file 'm4/ltoptions.m4'
    libtoolize: copying file 'm4/ltsugar.m4'
    libtoolize: copying file 'm4/ltversion.m4'
    libtoolize: copying file 'm4/lt~obsolete.m4'
    autoreconf: configure.ac: not using Intltool
    autoreconf: configure.ac: not using Gtkdoc
    autoreconf: running: aclocal --force -I m4
    autoreconf: running: /usr/bin/autoconf --force
    autoreconf: configure.ac: not using Autoheader
    autoreconf: running: automake --add-missing --copy --force-missing
    configure.ac:12: installing 'build-aux/compile'
    configure.ac:17: installing 'build-aux/config.guess'
    configure.ac:17: installing 'build-aux/config.sub'
    configure.ac:6: installing 'build-aux/install-sh'
    configure.ac:6: installing 'build-aux/missing'
    parallel-tests: installing 'build-aux/test-driver'
    EXAMPLES/MATRIX_MARKET/Makefile.am: installing 'build-aux/depcomp'
    autoreconf: Leaving directory '.'
    checking for a BSD-compatible install... /usr/bin/install -c
    checking whether build environment is sane... yes
    checking for a race-free mkdir -p... /usr/bin/mkdir -p
    checking for gawk... gawk
    checking whether make sets $(MAKE)... yes
    checking whether make supports nested variables... yes
    checking whether to enable maintainer-specific portions of Makefiles... no
    checking for gfortran... no
    checking for g95... no
    checking for xlf95... no
    checking for f95... no
    checking for fort... no
    checking for ifort... ifort
    checking whether the Fortran compiler works... no
    configure: error: in `/arpack-ng':
    configure: error: Fortran compiler cannot create executables
    See `config.log' for more details
    make: *** No rule to make target 'all'.  Stop.

If i try to write:

    ./configure --enable-icb
    cmake -D ICB=ON

It says:

    checking for a BSD-compatible install... /usr/bin/install -c
    checking whether build environment is sane... yes
    checking for a race-free mkdir -p... /usr/bin/mkdir -p
    checking for gawk... gawk
    checking whether make sets $(MAKE)... yes
    checking whether make supports nested variables... yes
    checking whether to enable maintainer-specific portions of Makefiles... no
    checking for gfortran... no
    checking for g95... no
    checking for xlf95... no
    checking for f95... no
    checking for fort... no
    checking for ifort... ifort
    checking whether the Fortran compiler works... no
    configure: error: in `/arpack-ng':
    configure: error: Fortran compiler cannot create executables
    See `config.log' for more details
    CMake Warning:
    No source or binary directory provided.  Both will be assumed to be the
    same as the current working directory, but note that this warning will
    become a fatal error in future CMake releases.

    -- The C compiler identification is GNU 11.4.0
    -- The Fortran compiler identification is IntelLLVM 2024.0.0
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Check for working C compiler: /usr/bin/cc - skipped
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Detecting Fortran compiler ABI info
    -- Detecting Fortran compiler ABI info - done
    -- Check for working Fortran compiler: /opt/intel/oneapi/compiler/2024.0/bin/ifx - skipped
    -- The CXX compiler identification is GNU 11.4.0
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Check for working CXX compiler: /usr/bin/c++ - skipped
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Fortran compiler does support iso_c_binding.
    -- Looking for Fortran sgemm
    -- Looking for Fortran sgemm - not found
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
    -- Found Threads: TRUE  
    -- Looking for Fortran sgemm
    -- Looking for Fortran sgemm - found
    -- Found BLAS: /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_lp64.so;/opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_thread.so;/opt/intel/oneapi/mkl/2024.0/lib/libmkl_core.so;/opt/intel/oneapi/compiler/2024.0/lib/libiomp5.so;-lm;-ldl  
    -- Looking for Fortran cheev
    -- Looking for Fortran cheev - found
    -- Found LAPACK: /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_lp64.so;/opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_thread.so;/opt/intel/oneapi/mkl/2024.0/lib/libmkl_core.so;/opt/intel/oneapi/compiler/2024.0/lib/libiomp5.so;-lm;-ldl;-lm;-ldl  
    -- Configuration summary for arpack-ng-3.9.1:
    -- prefix: /usr/local
    -- MPI: OFF (ICB provided )
    -- ICB: ON
    -- INTERFACE64: 0
    -- FC:      /opt/intel/oneapi/compiler/2024.0/bin/ifx
    -- FCFLAGS: -O3 -DMKL_ILP64 -I/usr/include/mkl
    -- CC:      /usr/bin/cc
    -- CCFLAGS: -O3 -DNDEBUG 
    -- CXX:      /usr/bin/c++
    -- CXXFLAGS: -O3 -DNDEBUG 
    -- BLAS:
        -- link:    /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_lp64.so
        -- link:    /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_thread.so
        -- link:    /opt/intel/oneapi/mkl/2024.0/lib/libmkl_core.so
        -- link:    /opt/intel/oneapi/compiler/2024.0/lib/libiomp5.so
        -- link:    -lm
        -- link:    -ldl
    -- LAPACK:
        -- link:    /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_lp64.so
        -- link:    /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_thread.so
        -- link:    /opt/intel/oneapi/mkl/2024.0/lib/libmkl_core.so
        -- link:    /opt/intel/oneapi/compiler/2024.0/lib/libiomp5.so
        -- link:    -lm
        -- link:    -ldl
        -- link:    -lm
        -- link:    -ldl
    -- Configuring done (1.8s)
    -- Generating done (0.0s)
    -- Build files have been written to: /arpack-ng

**Comment by fghoussen** [comment]
> When i try to use dsaupd_c and dseupd_c from arpack-ng

To use `*_c`, you need ICB: https://github.com/opencollab/arpack-ng#iso_c_binding-support.

Make sure you start from clean/empty directory, your env is correct (with intel: need to run sh script to set env before using compilers IIRC) and set ICB option at configure time. 

If using MKL: https://github.com/opencollab/arpack-ng#using-mkl-instead-of-blas--lapack

**Reply by Elil50** [reply]
I'm not really good at configuring etc (my second time. First one was just following blindfolded the instructions I wrote).

I will now redownload all the pack in a different directory and link that one.

What are the exact steps I need to do?

I see the 2 steps you wrote, but trying them lead to an issue and I don't know if it is related to having already installed and configured everything. Expecially for the fact that it says fortran compiler doesn't work.

Thanks for the hard work.

**Reply by fghoussen** [reply]
> I'm not really good at configuring 

Can't help from here...

> What are the exact steps I need to do?

- Make sure your env is OK (compiler, SDK install). Start with gnu/clang compiler first (intel is more complex).
- If using MKL, make sure MKL is installed properly and follow __exactly__ intel's doc to link. Start without MKL first (intel is more complex): MKL is a mess even for intel people: that's why they provide the intel advisor...

CI runs ICB with MKL: means CMake works. Problems are generally always caused by bad env. Can't help from here sorry

> trying them lead to an issue and I don't know

No trace... No clue... Check logs

**Reply by Elil50** [reply]
It's not an issue on configuring mkl and its enviroment, as it already works in both c++ and fortran programs.

These are the steps performed:

    git clone https://github.com/opencollab/arpack-ng.git
    cd ./arpack-ng

    mkdir build
    cd build

    cmake -D ICB=ON -D INTERFACE64=ON -D ITF64SUFFIX="ILP64" -D BUILD_SHARED_LIBS=ON -D BLA_VENDOR=Intel10_64ilp ..

    make
    sudo make install


This is my cmake:

    # Home made C++ libraries
    add_library(_quantum_Ising_mkl SHARED)
    target_sources(_quantum_Ising_mkl PRIVATE
        ${executable_and_libraries}/src_homemade/quantum_Ising_mkl.cpp
    )
    target_link_libraries(_quantum_Ising_mkl PUBLIC
        "-L${MKLROOT}/lib -lmkl_intel_ilp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl"
    )
    target_include_directories(_quantum_Ising_mkl PUBLIC
        ${executable_and_libraries}/arpack-ng/SRC
        ${executable_and_libraries}/arpack-ng/build
        ${executable_and_libraries}/arpack-ng/ICB
        ${MKLROOT}/include
    )
    set_target_properties(_quantum_Ising_mkl PROPERTIES
        COMPILE_FLAGS  "-DMKL_ILP64 -Ofast -march=native"
    )

This is the result after running Cmakelist.txt

    Fortran_compiler: ifort
    CXX_compiler: icpx
    -- MKL_VERSION: 2024.0.0
    -- MKL_ROOT: /opt/intel/oneapi/mkl/2024.0
    -- MKL_SYCL_ARCH: intel64
    -- MKL_ARCH: intel64
    -- MKL_SYCL_LINK: static
    -- MKL_LINK: static
    -- MKL_SYCL_INTERFACE_FULL: intel_ilp64
    -- MKL_INTERFACE_FULL: intel_ilp64
    -- MKL_SYCL_THREADING: sequential
    -- MKL_THREADING: sequential
    -- Found /opt/intel/oneapi/mkl/2024.0/lib/libmkl_intel_ilp64.a
    -- Found /opt/intel/oneapi/mkl/2024.0/lib/libmkl_sequential.a
    -- Found /opt/intel/oneapi/mkl/2024.0/lib/libmkl_core.a
    -- Found /opt/intel/oneapi/mkl/2024.0/lib/libmkl_sycl.a
    -- Configuring done (0.0s)
    -- Generating done (0.0s)
    -- Build files have been written to: /executable_and_libraries/build
    [ 50%] Built target _Quantum_Ising_1D_double_multithreadable
    [ 75%] Building CXX object CMakeFiles/_quantum_Ising_mkl.dir/src_homemade/quantum_Ising_mkl.cpp.o
    [100%] Linking CXX shared library lib_quantum_Ising_mkl.so
    [100%] Built target _quantum_Ising_mkl


This is the error I get when calling the function (I call the C++ library I made (which calls dseupd) from a python file (this worked fine with every other library I wrote))

    File "/usr/lib/python3.10/ctypes/__init__.py", line 374, in __init__
      self._handle = _dlopen(self._name, mode)
    OSError: ../build/lib_quantum_Ising_mkl.so: undefined symbol: dseupd_c

That's the same issue.

In my C++ library I included:

    #include "arpack.h"
    #include "arpackdef.h"
    #include "mkl_spblas.h"

I first call dsaupd:

    dsaupd_c (&ido, Bmat, tot_Hilbert_elem, which_eval, nev, tol, v0, ncv,
            lancz_basis_matrix, ldv, iparam, ipntr, workd, workl, length_workl, &info );

Then call dseupd_c:

    dseupd_c ( evec_compute_flag, howmny, select, eval_array, lancz_basis_matrix, ldv, sigma, 
    Bmat, tot_Hilbert_elem, which_eval, nev, tol, v0, ncv, lancz_basis_matrix, ldv, 
    iparam, ipntr, workd, workl, length_workl, &ierr);

It's quite funnny as now vsc no longer sees these function arguments as int64_t, but as int.
However if i rewrite them all as int and compile, it gives error, as they need int64_t.

If i leave them as int64_t it does not find dseupd_c, which is found by vsc (ctrl+left click) and is the second function I call.

**Reply by Elil50** [reply]
![Screenshot from 2024-01-21 23-21-43](https://github.com/opencollab/arpack-ng/assets/103074892/bd8b448d-9824-463d-a6f5-733bd4c85361)

I just found this inside arpack.h, following those passages, so surely it's a build issue

**Comment by sylvestre** [comment]
please don't share screenshoots, they are terrible for accessibility and search
thanks

**Reply by Elil50** [reply]
To sum up, vsc suggestion can't find the functions which are linked to the ones written in arpack.h. That's a build issue. Can you try the passages I made and tell me if you incur in the same problem? Have I missed an important step in the configuration process?

**Reply by fghoussen** [reply]
> That's a build issue. 

It's  not a build issue but likely a link issue (export symbols) at your side.

**Reply by Elil50** [reply]
Yes, you could be right. Then did I miss some libraries or directories to link in the cmakelist I wrote?

**Reply by fghoussen** [reply]
Very difficult to say / help from here...
Does the test suite runs OK: is `make test`OK ?
Which compiler/linker do you use?
What is your OS: unix? windows?

**Reply by fghoussen** [reply]
>  did I miss some libraries or directories to lin

Example here: https://github.com/opencollab/arpack-ng/blob/ba440d13229972b161de37d08dd61bd5a0aa07f2/cmake/tstCMakeInstall.sh.in#L72

**Reply by Elil50** [reply]
I would try this afternoon (7 hours ahead of now). Thanks. I use Linux. If I remember well the test gave same issue in the terminal when I compiled (not runned) with ICB. Is it supposed to happen? If not I would try all the steps again and paste the output

**Reply by fghoussen** [reply]
> I use Linux.

Should work on Linux. I thought at an eventual link problem on Windows: AFAIK, ICB has not been tested on Windows

**Comment by Elil50** [comment]
Following these steps:
        
        git clone https://github.com/opencollab/arpack-ng.git
        cd ./arpack-ng
        
        mkdir build
        cd build
        
        cmake -D ICB=ON -D INTERFACE64=ON -D ITF64SUFFIX="ILP64" -D BUILD_SHARED_LIBS=ON -D BLA_VENDOR=Intel10_64ilp ..
        
        make
        sudo make install

This is the output of make:

    [ 97%] Building C object CMakeFiles/icb_arpack_c.dir/TESTS/icb_arpack_c.c.o
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c: In function ‘ds’:
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:69:40: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    69 |     printf("Error in saupd: iparam[4] %d, nev %d, info %d\n", iparam[4], nev, info);
        |                                       ~^                      ~~~~~~~~~
        |                                        |                            |
        |                                        int                          int64_t {aka long int}
        |                                       %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:69:48: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    69 |     printf("Error in saupd: iparam[4] %d, nev %d, info %d\n", iparam[4], nev, info);
        |                                               ~^                         ~~~
        |                                                |                         |
        |                                                int                       int64_t {aka long int}
        |                                               %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:69:57: warning: format ‘%d’ expects argument of type ‘int’, but argument 4 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    69 |     printf("Error in saupd: iparam[4] %d, nev %d, info %d\n", iparam[4], nev, info);
        |                                                        ~^                     ~~~~
        |                                                         |                     |
        |                                                         int                   int64_t {aka long int}
        |                                                        %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:76:35: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    76 |     printf("Error in seupd: info %d\n", info);
        |                                  ~^     ~~~~
        |                                   |     |
        |                                   int   int64_t {aka long int}
        |                                  %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c: In function ‘zn’:
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:140:40: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    140 |     printf("Error in naupd: iparam[4] %d, nev %d, info %d\n", iparam[4], nev, info);
        |                                       ~^                      ~~~~~~~~~
        |                                        |                            |
        |                                        int                          int64_t {aka long int}
        |                                       %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:140:48: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    140 |     printf("Error in naupd: iparam[4] %d, nev %d, info %d\n", iparam[4], nev, info);
        |                                               ~^                         ~~~
        |                                                |                         |
        |                                                int                       int64_t {aka long int}
        |                                               %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:140:57: warning: format ‘%d’ expects argument of type ‘int’, but argument 4 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    140 |     printf("Error in naupd: iparam[4] %d, nev %d, info %d\n", iparam[4], nev, info);
        |                                                        ~^                     ~~~~
        |                                                         |                     |
        |                                                         int                   int64_t {aka long int}
        |                                                        %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:147:35: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    147 |     printf("Error in neupd: info %d\n", info);
        |                                  ~^     ~~~~
        |                                   |     |
        |                                   int   int64_t {aka long int}
        |                                  %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c: In function ‘main’:
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:182:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    182 |   printf("Timers : nopx %d, tmvopx %f - nbx %d, tmvbx %f\n", nopx_c, tmvopx_c,
        |                         ~^                                   ~~~~~~
        |                          |                                   |
        |                          int                                 int64_t {aka long int}
        |                         %ld
    /home/matteobacci/Desktop/Programmi/MODULO_02/executable_and_libraries/arpack-ng/TESTS/icb_arpack_c.c:182:46: warning: format ‘%d’ expects argument of type ‘int’, but argument 4 has type ‘int64_t’ {aka ‘long int’} [-Wformat=]
    182 |   printf("Timers : nopx %d, tmvopx %f - nbx %d, tmvbx %f\n", nopx_c, tmvopx_c,
        |                                             ~^
        |                                              |
        |                                              int
        |                                             %ld
    183 |          nbx_c, tmvbx_c);
        |          ~~~~~                                
        |          |
        |          int64_t {aka long int}
    [ 98%] Linking C executable TESTS/icb_arpack_c
    [ 98%] Built target icb_arpack_c
    [ 99%] Building CXX object CMakeFiles/icb_arpack_cpp.dir/TESTS/icb_arpack_cpp.cpp.o
    [100%] Linking CXX executable TESTS/icb_arpack_cpp

Dunno if it's an issue, even though it's not about linked libraries, i still don't know if t's ok to run tests with it.

About cmakelist.txt, I included:

    add_library(_quantum_Ising_mkl SHARED)
    target_sources(_quantum_Ising_mkl PRIVATE
        ${executable_and_libraries}/src_homemade/quantum_Ising_mkl.cpp
    )
    target_link_libraries(_quantum_Ising_mkl PUBLIC
        "-L${MKLROOT}/lib -lmkl_intel_ilp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl"
    )
    target_include_directories(_quantum_Ising_mkl PUBLIC
        ${executable_and_libraries}/arpack-ng/SRC
        ${executable_and_libraries}/arpack-ng/build
        ${executable_and_libraries}/arpack-ng/ICB
        ${MKLROOT}/include
    )
    set_target_properties(_quantum_Ising_mkl PROPERTIES
        COMPILE_FLAGS  "-DMKL_ILP64 -Ofast -march=native"
    )

From the line you linked I see:

    echo "target_include_directories(icb_arpack_c INTERFACE BLAS::BLAS)"     >> CMakeLists.txt
    echo "target_link_libraries(icb_arpack_c BLAS::BLAS)"                    >> CMakeLists.txt
    echo "target_include_directories(icb_arpack_c INTERFACE LAPACK::LAPACK)" >> CMakeLists.txt
    echo "target_link_libraries(icb_arpack_c LAPACK::LAPACK)"                >> CMakeLists.txt
    echo "target_include_directories(icb_arpack_c INTERFACE ARPACK::ARPACK)" >> CMakeLists.txt
    echo "target_link_libraries(icb_arpack_c ARPACK::ARPACK)"                >> CMakeLists.txt          

If I try to paste stuff like that, i need to know where it is. It really reminds me on how i link on fortran:

    add_executable(main_program_fortran)
    target_sources(main_program_fortran PUBLIC
        ${executable_and_libraries}/bin/main_program_fortran.f90
    )
    target_link_libraries(main_program_fortran PUBLIC 

        ${executable_and_libraries}/src_external/arpack-ng/build/libarpackILP64.a
        ${executable_and_libraries}/src_external/mkl_spblas.o
        ${MKLROOT}/lib/intel64/libmkl_blas95_ilp64.a
        ${MKLROOT}/lib/intel64/libmkl_lapack95_ilp64.a
        "-Wl,--start-group ${MKLROOT}/lib/libmkl_intel_ilp64.a ${MKLROOT}/lib/libmkl_sequential.a ${MKLROOT}/lib/libmkl_core.a -Wl,--end-group -lpthread -lm -ldl")
    target_include_directories(main_program_fortran PUBLIC
        ${MKLROOT}/include/mkl/intel64/ilp64
        ${MKLROOT}/include
    )
    set_target_properties(main_program_fortran PROPERTIES
        COMPILE_FLAGS "-Ofast -i8 -r8 -static"
    )

However, links are differents for dynamic libraries in mkl (even though that's not an issue of mkl)
What I see is that you link ARPACK::ARPACK, but I don't know where it is and that's probably what I'm missing.

In fortran I was building static libraries too, such as:

    /arpack-ng/build/libarpackILP64.a

So, if I try:

    cd ../
    rm -r -f build
    mkdir build
    cd build
    cmake -D ICB=ON -D INTERFACE64=ON -D ITF64SUFFIX="ILP64" -D BUILD_SHARED_LIBS=OFF -D BLA_VENDOR=Intel10_64ilp ..
    make
    sudo make install

It says:

    [ 86%] Linking C executable TESTS/bug_1315_single
    /usr/bin/ld: skipping incompatible /usr/lib/i386-linux-gnu/libgcc_s.so.1 when searching for libgcc_s.so.1
    /usr/bin/ld: skipping incompatible /usr/lib/i386-linux-gnu/libgcc_s.so.1 when searching for libgcc_s.so.1
    /usr/bin/ld: libarpackILP64.a(icbasn.F90.o): relocation R_X86_64_32 against `.bss' can not be used when making a PIE object; recompile with -fPIE
    /usr/bin/ld: failed to set dynamic section sizes: bad value
    collect2: error: ld returned 1 exit status
    make[2]: *** [CMakeFiles/bug_1315_single.dir/build.make:102: TESTS/bug_1315_single] Error 1
    make[1]: *** [CMakeFiles/Makefile2:155: CMakeFiles/bug_1315_single.dir/all] Error 2
    make: *** [Makefile:166: all] Error 2
    [sudo] password for: 
    [ 81%] Built target arpack
    [ 84%] Built target dnsimp_test
    [ 85%] Linking C executable TESTS/bug_1315_single
    /usr/bin/ld: cannot find -lifport: No such file or directory
    /usr/bin/ld: cannot find -lifcoremt: No such file or directory
    /usr/bin/ld: cannot find -limf: No such file or directory
    /usr/bin/ld: cannot find -lsvml: No such file or directory
    collect2: error: ld returned 1 exit status
    make[2]: *** [CMakeFiles/bug_1315_single.dir/build.make:102: TESTS/bug_1315_single] Error 1
    make[1]: *** [CMakeFiles/Makefile2:155: CMakeFiles/bug_1315_single.dir/all] Error 2
    make: *** [Makefile:166: all] Error 2

And if I add that library to the cmake and run it:

    /usr/bin/ld: /arpack-ng/build/libarpackILP64.a(icbads.F90.o): relocation R_X86_64_32 against `.bss' can not be used when making a shared object; recompile with -fPIC
    /usr/bin/ld: failed to set dynamic section sizes: bad value
    icpx: error: linker command failed with exit code 1 (use -v to see invocation)
    gmake[2]: *** [CMakeFiles/_quantum_Ising_mkl.dir/build.make:98: lib_quantum_Ising_mkl.so] Error 1
    gmake[1]: *** [CMakeFiles/Makefile2:111: CMakeFiles/_quantum_Ising_mkl.dir/all] Error 2
    gmake: *** [Makefile:91: all] Error 2

**Reply by fghoussen** [reply]
No time to analyse this.

My advice: you seem to mix everything (ICB + ILP64 + python + icc/ifort), stop it! Start simple and minimal: gcc + ICB? Is it enough/OK (is ILP64 really needed)? If yes add what you need on top of it (each added feature will come with a price/complexity...)

Keep things simple as most as possible:

> target_link_libraries(_quantum_Ising_mkl PUBLIC
    "-L${MKLROOT}/lib -lmkl_intel_ilp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl"
)
target_include_directories(_quantum_Ising_mkl PUBLIC
    ${executable_and_libraries}/arpack-ng/SRC
    ${executable_and_libraries}/arpack-ng/build
    ${executable_and_libraries}/arpack-ng/ICB
    ${MKLROOT}/include
)

is overkill, as this

> target_include_directories(icb_arpack_c INTERFACE ARPACK::ARPACK)
target_link_libraries(icb_arpack_c ARPACK::ARPACK)

Should be enough: this is the minimum you need. If all is OK, `BLAS::BLAS` and `LAPACK::LAPACK` will point to the correct implementation (MKL or any other one). Do not set thing "by hand" unless you really need to set things manually (and then you generally need to set everything by hand). So, don't do that?!

> target_link_libraries(main_program_fortran PUBLIC 
    ${executable_and_libraries}/src_external/arpack-ng/build/libarpackILP64.a
    ${executable_and_libraries}/src_external/mkl_spblas.o
    ${MKLROOT}/lib/intel64/libmkl_blas95_ilp64.a
    ${MKLROOT}/lib/intel64/libmkl_lapack95_ilp64.a
    "-Wl,--start-group ${MKLROOT}/lib/libmkl_intel_ilp64.a ${MKLROOT}/lib/libmkl_sequential.a ${MKLROOT}/lib/libmkl_core.a -Wl,--end-group -lpthread -lm -ldl")
target_include_directories(main_program_fortran PUBLIC
    ${MKLROOT}/include/mkl/intel64/ilp64
    ${MKLROOT}/include
)

There is no chance this could work... Pass this flags to LDFLAGS and friends if, and only if,  the intel link advisor tells you to do so.

I recommend not using MKL for a start. If this is really necessary, using MKL: https://github.com/opencollab/arpack-ng#using-mkl-instead-of-blas--lapack
- Either `BLA_VENDOR=Intel10_64lp` is OK and you are fine: don't "set" anything manually
- Or, then, you'll have to "set" everything manually (= drop BLA_VENDOR - do NOT mix "manual" with "automatic" settings as you'll have 100% to get something that has no chance to work - go one way OR the other): use the "intel link advisor" following https://github.com/opencollab/arpack-ng/blob/ba440d13229972b161de37d08dd61bd5a0aa07f2/.github/workflows/jobs.yml#L162 that you'll __need to adapt__ (in the CI, we use MKL to test ILP64 purposely as it is the only BLAS/LAPACK lib that comes with ILP64 flavour and available in open-source - drop ILP64 if you don't need it)

Do ONLY minimal changes and settings: it should work. Start simple. don't mix manual/automatic settings. Add complexity only if needed.

The test suite should work: don't go further until it doesn't.

Can't afford spending more time on that sorry.

**Comment by Elil50** [comment, marked as answer]
Ok, so in the end I just needed to add this library.:

      /arpack-ng/build/lib/libarpackILP64.so

It seems developers didn't spot it. I went through all automatically generated makefiles to find this piece was missing
